monitoring/system_pods.go

```diff
@@ -0,0 +1,218 @@
+		errors = append(errors, trace.BadParameter("node name must be provided"))
+
+	}
+	if r.KubeConfig == nil {
+		errors = append(errors, trace.BadParameter("kubernetes access config must be provided"))
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// systemPodsChecker verifies system pods are operational.
+type systemPodsChecker struct {
+	// SystemPodsConfig specifies checker configuration values.
+	SystemPodsConfig
+}
+
+// NewSystemPodsChecker returns a new system pods checker.
+func NewSystemPodsChecker(config SystemPodsConfig) (*systemPodsChecker, error) {
+	if err := config.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &systemPodsChecker{
+		SystemPodsConfig: config,
+	}, nil
+}
+
+// Name returns this checker name
+// Implements health.Checker
+func (r *systemPodsChecker) Name() string {
+	return systemPodsCheckerID
+}
+
+// Check verifies that all system pods are operational.
+// Implements health.Checker
+func (r *systemPodsChecker) Check(ctx context.Context, reporter health.Reporter) {
+	err := r.check(ctx, reporter)
+	if err != nil {
+		log.WithError(err).Warn("Failed to verify critical system pods")
+		// TODO: set probe to critical
+		reporter.Add(&pb.Probe{
+			Checker:  r.Name(),
+			Detail:   "failed to verify critical system pods",
+			Error:    trace.UserMessage(err),
+			Status:   pb.Probe_Failed,
+			Severity: pb.Probe_Warning,
+		})
+		return
+	}
+	if reporter.NumProbes() == 0 {
+		reporter.Add(NewSuccessProbe(r.Name()))
+	}
+}
+
+func (r *systemPodsChecker) check(ctx context.Context, reporter health.Reporter) error {
+	pods, err := r.getPods()
+	if trace.IsNotFound(err) {
+		log.Debug("No critical system pods found.")
+		return nil // system pods were not found, log and treat gracefully
+	}
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	r.verifyPods(pods, reporter)
+	return nil
+}
+
+// getPods returns a list of the local pods that have the `critical` label.
+func (r *systemPodsChecker) getPods() ([]corev1.Pod, error) {
+	opts := metav1.ListOptions{
+		LabelSelector: systemPodsSelector.String(),
+		FieldSelector: fields.OneTermEqualSelector("spec.nodeName", r.NodeName).String(),
+	}
+	pods, err := r.Client.CoreV1().Pods("").List(opts)
+	if err != nil {
+		return nil, utils.ConvertError(err)
+	}
+
+	return pods.Items, nil
+}
+
+// verifyPods verifies the pods are in a valid state. Reports a failed probe for
+// any pods that are in an invalid state.
+func (r *systemPodsChecker) verifyPods(pods []corev1.Pod, reporter health.Reporter) {
+	for _, pod := range pods {
+		if err := verifyPodStatus(pod.Status); err != nil {
+			reporter.Add(systemPodsFailureProbe(r.Name(), pod.Name, pod.Namespace))
+		}
+	}
+}
+
+// verifyPodStatus verifies the status phase and conditions.
+func verifyPodStatus(status corev1.PodStatus) error {
+	// https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase
+	switch status.Phase {
+	case corev1.PodPending, corev1.PodSucceeded:
```

Comment:
Regarding Pending status - do you know if there's a way to distinguish b/w when the pod is just pending to be scheduled versus when it's forever pending due to, for example, being unable to attach a persistent volume? While the former is ok, the latter case might be worth highlighting. Not sure there's a way to tell b/w the two though, I would just try it out and see.

Reply:
So I'm not sure this checker can even verify either case. With the current implementation, if a pod is not yet scheduled to a node, the pod will not be verified. It will be filtered when we are querying local pods.

If we wanted to check pods that are unscheduled, we would need to modify the `getNodes` function or separately query pods that are not local to the node and verify them. All nodes would have to verify the same pods and we'd end up with duplicate probes on all nodes. 

If this is worth including, we could just have generic message saying there are some critical probes that have not been scheduled. Something like:

```
Cluster nodes:
    Masters:
        * node-1 / 172.28.128.101 / node
            Status:             healthy
            [!]                 One more more critical system pods have not been scheduled
            Remote access:      online
        * node-3 / 172.28.128.103 / node
            Status:             healthy
            [!]                 One more more critical system pods have not been scheduled
            Remote access:      online
        * node-2 / 172.28.128.102 / node
            Status:             healthy
            [!]                 One more more critical system pods have not been scheduled
            Remote access:      online
```

Also not sure if we would be able to distinguish the two scenarios. I don't think there is any field that changes if the pod has been continually failing to be scheduled?
```
Name:           task-pv-pod-7d898d594f-dq5bz
Namespace:      default
Priority:       0
Node:           <none>
Labels:         app=pv
                gravitational.io/critical-pod=true
                pod-template-hash=7d898d594f
Annotations:    kubernetes.io/psp: coredns
                seccomp.security.alpha.kubernetes.io/pod: docker/default
Status:         Pending
IP:
IPs:            <none>
Controlled By:  ReplicaSet/task-pv-pod-7d898d594f
Containers:
  task-pv-container:
    Image:        nginx
    Port:         80/TCP
    Host Port:    0/TCP
    Environment:  <none>
    Mounts:
      /usr/share/nginx/html from task-pv-storage (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-kpn6m (ro)
Conditions:
  Type           Status
  PodScheduled   False
Volumes:
  task-pv-storage:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  task-pv-claim
    ReadOnly:   false
  default-token-kpn6m:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-kpn6m
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason            Age        From               Message
  ----     ------            ----       ----               -------
  Warning  FailedScheduling  <unknown>  default-scheduler  persistentvolumeclaim "task-pv-claim" not found
  Warning  FailedScheduling  <unknown>  default-scheduler  persistentvolumeclaim "task-pv-claim" not found
```